src/freenet/clients/http/FirstTimeWizardNewToadlet.java

```diff
@@ -0,0 +1,329 @@
+            HashMap<String, Object> model = new HashMap<>();
+            model.put("knowSomeone", knowSomeone.length() > 0 ? "checked" : "");
+            model.put("connectToStrangers", connectToStrangers.length() > 0 ? "checked" : "");
+            model.put("haveMonthlyLimit", haveMonthlyLimit.length() > 0 ? "checked" : "");
+            model.put("downloadLimit", downloadLimit);
+            model.put("uploadLimit", uploadLimit);
+            model.put("bandwidthMonthlyLimit", bandwidthMonthlyLimit);
+            model.put("minBandwidthMonthlyLimit", String.format("%.2f", BandwidthLimit.minMonthlyLimit));
+            model.put("storageLimit", storageLimit);
+            if (isPasswordEmpty) {
+                model.put("setPassword", setPassword.length() > 0 ? "checked" : "");
+            }
+            model.put("isPasswordEmpty", isPasswordEmpty);
+
+            if (downloadLimitDetected == null || uploadLimitDetected == null) {
+                detectBandwidthLimit();
+            }
+            model.put("downloadLimitDetected", downloadLimitDetected != null ? downloadLimitDetected : l10n("bandwidthCommonInternetConnectionSpeedsDetectedUnavailable"));
+            model.put("uploadLimitDetected", uploadLimitDetected != null ? uploadLimitDetected : l10n("bandwidthCommonInternetConnectionSpeedsDetectedUnavailable"));
+
+            model.put("errors", errors);
+
+            return model;
+        }
+
+        private void save() {
+            if (knowSomeone.isEmpty()) {
+                // Opennet
+                core.node.securityLevels.setThreatLevel(SecurityLevels.NETWORK_THREAT_LEVEL.LOW);
+            } else {
+                if (connectToStrangers.isEmpty()) {
+                    // Darknet
+                    core.node.securityLevels.setThreatLevel(SecurityLevels.NETWORK_THREAT_LEVEL.HIGH);
+                } else {
+                    // Opennet + Darknet
+                    core.node.securityLevels.setThreatLevel(SecurityLevels.NETWORK_THREAT_LEVEL.NORMAL);
+                }
+            }
+
+            try {
+                if (haveMonthlyLimit.isEmpty()) { // save download & uploadLimit
+                    config.get("node").set("inputBandwidthLimit", downloadLimit + "KiB");
+                    config.get("node").set("outputBandwidthLimit", uploadLimit + "KiB");
+                } else { // save bandwidthMonthlyLimit
+                    BandwidthLimit bandwidth = new BandwidthLimit(Fields.parseLong(bandwidthMonthlyLimit + "GiB"));
+                    config.get("node").set("inputBandwidthLimit", Long.toString(bandwidth.downBytes));
+                    config.get("node").set("outputBandwidthLimit", Long.toString(bandwidth.upBytes));
+                }
+            } catch (ConfigException e) {
+                Logger.error(this, "Should not happen, please report! " + e, e);
+            }
+
+            DATASTORE_SIZE.setDatastoreSize(storageLimit + "GiB", config, this);
+
+            // TODO: not sure
```

Comment:
Not sure about what?

This block looks like it might be incorrect? `isPasswordEmpty` - which looks like it's whether the physical threat level results in a password being set - is checked first, but then there's a branch that sets the master password. Does this only call `core.node.setMasterPassword(password, true)` when it's not supposed to?

Thoughts on renaming `isPasswordEmpty` to something like `passwordRequested`? I was confused by checking both `isPasswordEmpty` and `setPassword.isEmpty()`. Am I missing something or is `setPassword` non-empty if and only if `isPasswordEmpty` is true? Do we need both variables to exist? Are they separate because `isPasswordEmpty` is only set on GET? What's the use of the value being set only on GET? Is it an initial setting for password-ness within the page?

Would it make sense for this block to be something like:

```
                try {
                    if (passwordRequested) {
                        core.node.securityLevels.setThreatLevel(SecurityLevels.PHYSICAL_THREAT_LEVEL.HIGH);
                        core.node.setMasterPassword(password, true);
                    } else {
                        core.node.securityLevels.setThreatLevel(SecurityLevels.PHYSICAL_THREAT_LEVEL.NORMAL);
                        core.node.setMasterPassword("", true);
                    }
                } catch (Node.AlreadySetPasswordException | MasterKeysWrongPasswordException | MasterKeysFileSizeException | IOException e) {
                    Logger.error(this, "Should not happen, please report! " + e, e);
                }
```

Reply:
That’s a good point … your suggestion sounds good (and I guess that was the "not sure" in the TODO), but I’ll have to test this. Thank you!

Reply:
I adjusted this and also added a few additional fixes.